nats_test.go

```diff
@@ -2540,3 +2541,76 @@ func TestLameDuckMode(t *testing.T) {
 	}
 	wg.Wait()
 }
+
+func TestJSSubscribe(t *testing.T) {
+	srv, _, _, nc := startJetStream(t)
+	defer os.RemoveAll(srv.JetStreamConfig().StoreDir)
+	defer srv.Shutdown()
+	defer nc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 250*time.Millisecond)
+	defer cancel()
+
+	seen := 0
+	cb := func(m *Msg) {
+		m.Ack()
+		seen++
+		if seen == 20 {
+			cancel()
+		}
+	}
+
+	cfg := &ConsumerConfig{
+		Durable:        "nats",
+		DeliverSubject: NewInbox(),
+		DeliverPolicy:  DeliverAll,
+		AckPolicy:      AckExplicit,
+		AckWait:        5 * time.Second,
+		ReplayPolicy:   ReplayInstant,
+	}
+
+	sub, err := nc.Subscribe("", cb, Consumer("TEST", cfg))
+	if err != nil {
+		t.Fatalf("create failed: %s", err)
+	}
+	defer sub.Unsubscribe()
+
+	if sub.ConsumerConfig.Durable != "nats" {
+		t.Fatalf("got wrong durable: %q", sub.ConsumerConfig.Durable)
+	}
+	if sub.Subject != cfg.DeliverSubject {
+		t.Fatalf("subscriber not subscribed to the delivery subject")
+	}
+
+	<-ctx.Done()
+
+	if seen != 20 {
+		t.Fatalf("Expected 20 messages got %d", seen)
+	}
+
+	// should fail to update if existing subscribe is active (server error)
+	cfg.DeliverSubject = NewInbox()
+	_, err = nc.Subscribe("", cb, Consumer("TEST", cfg))
+	if err == nil {
+		t.Fatal("create succeeded, expected error")
+	}
+	if err.Error() != "nats: JetStream consumer creation failed: consumer already exists" {
+		t.Fatalf("expected already exist error got %q", err.Error())
+	}
+
+	sub.Unsubscribe()
+
+	// should be able to create an ephemeral which requires interest to exist first
+	// before creating the ephemeral
+	cfg.Durable = ""
+	sub, err = nc.Subscribe("test", cb, Consumer("TEST", cfg))
```

Comment:
For instance, here we have a `cb` so we know we need push, we can do the inbox thing for the user and not make them do all the boilerplate from above.

Reply:
Would leave as is for the moment until we know how pull handling will actually look, easier to add the bit of magic than take it away if it seems weird later